pkg/webhook/core.oam.dev/v1alpha2/traitdefinition/validating_handler.go

```diff
@@ -0,0 +1,219 @@
+package traitdefinition
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	admissionv1beta1 "k8s.io/api/admission/v1beta1"
+	"k8s.io/klog"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/runtime/inject"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+
+	"cuelang.org/go/cue"
+	"cuelang.org/go/cue/build"
+	"github.com/pkg/errors"
+
+	"github.com/oam-dev/kubevela/apis/core.oam.dev/v1alpha2"
+	"github.com/oam-dev/kubevela/pkg/oam/discoverymapper"
+)
+
+const (
+	errDefinitionReferenceRequired = "definition reference is required"
+	errGVKRequired                 = "output must have GVK in CUE template"
+)
+
+const (
+	apiVersionFieldName = "apiVersion"
+	kindFieldName       = "kind"
+)
+
+var traitDefGVR = v1alpha2.SchemeGroupVersion.WithResource("traitdefinitions")
+
+// ValidatingHandler handles validation of trait definition
+type ValidatingHandler struct {
+	Client client.Client
+	Mapper discoverymapper.DiscoveryMapper
+
+	// Decoder decodes object
+	Decoder *admission.Decoder
+	// Validators validate objects
+	Validators []TraitDefValidator
+}
+
+// TraitDefValidator validate trait definition
+type TraitDefValidator interface {
+	Validate(context.Context, v1alpha2.TraitDefinition) error
+}
+
+// TraitDefValidatorFn implements TraitDefValidator
+type TraitDefValidatorFn func(context.Context, v1alpha2.TraitDefinition) error
+
+// Validate implements TraitDefValidator method
+func (fn TraitDefValidatorFn) Validate(ctx context.Context, td v1alpha2.TraitDefinition) error {
+	return fn(ctx, td)
+}
+
+var _ admission.Handler = &ValidatingHandler{}
+
+// Handle validate trait definition
+func (h *ValidatingHandler) Handle(ctx context.Context, req admission.Request) admission.Response {
+	obj := &v1alpha2.TraitDefinition{}
+	if req.Resource.String() != traitDefGVR.String() {
+		return admission.Errored(http.StatusBadRequest, fmt.Errorf("expect resource to be %s", traitDefGVR))
+	}
+	switch req.Operation { //nolint:exhaustive
+	case admissionv1beta1.Delete:
+		if len(req.OldObject.Raw) != 0 {
+			if err := h.Decoder.DecodeRaw(req.OldObject, obj); err != nil {
+				return admission.Errored(http.StatusBadRequest, err)
+			}
+		} else {
+			klog.Info("deleting Trait Definition", req.Name)
+		}
+	default:
```

Comment:
```suggestion
	case admissionv1beta1.Create, admissionv1beta1.Update:
```

Please make it clean, you should not include `Option` event here.